File: hphp/hack/src/utils/rust/pos/file_pos_large.rs
// Copyright (c) Facebook, Inc. and its affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the "hack" directory of this source tree.

use ocamlrep::FromOcamlRep;
use ocamlrep::FromOcamlRepIn;
use ocamlrep::ToOcamlRep;
use serde::Deserialize;
use serde::Serialize;

use crate::file_pos::FilePos;
use crate::file_pos_small::FilePosSmall;

#[derive(Copy, Clone, Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
pub struct FilePosLarge {
    /// line number. Starts at 1.
    lnum: usize,
    /// character number of the beginning of line of this position.
    /// The column number is therefore offset - bol
    /// Starts at 0
    bol: usize,
    /// character offset from the beginning of the file. Starts at 0.
    offset: usize,
}
arena_deserializer::impl_deserialize_in_arena!(FilePosLarge);

impl arena_trait::TrivialDrop for FilePosLarge {}

const DUMMY: FilePosLarge = FilePosLarge {
    lnum: 0,
    bol: 0,
    offset: usize::max_value(),
};

impl FilePosLarge {
    #[inline]
    pub const fn make_dummy() -> Self {
        DUMMY
    }

    #[inline]
    pub fn is_dummy(self) -> bool {
        self == DUMMY
    }

    #[inline]
    pub const fn beg_of_file(self) -> Self {
        FilePosLarge {
            lnum: 1,
            bol: 0,
            offset: 0,
        }
    }

    // constructors

    #[inline]
    pub const fn from_line_column_offset(line: usize, column: usize, offset: usize) -> Self {
        FilePosLarge {
            lnum: line,
            bol: offset - column,
            offset,
        }
    }

    #[inline]
    pub const fn from_lnum_bol_offset(lnum: usize, bol: usize, offset: usize) -> Self {
        FilePosLarge { lnum, bol, offset }
    }

    // accessors

    #[inline]
    pub const fn line(self) -> usize {
        self.lnum
    }

    #[inline]
    pub const fn column(self) -> usize {
        self.offset - self.bol
    }

    #[inline]
    pub const fn beg_of_line(self) -> usize {
        self.bol
    }

    #[inline]
    pub const fn with_column(self, col: usize) -> Self {
        FilePosLarge {
            lnum: self.lnum,
            bol: self.bol,
            offset: self.bol + col,
        }
    }

    #[inline]
    pub const fn line_beg(self) -> (usize, usize) {
        (self.lnum, self.bol)
    }

    #[inline]
    pub const fn line_column(self) -> (usize, usize) {
        (self.lnum, self.offset - self.bol)
    }

    #[inline]
    pub const fn line_column_offset(self) -> (usize, usize, usize) {
        (self.lnum, self.offset - self.bol, self.offset)
    }

    #[inline]
    pub const fn line_beg_offset(self) -> (usize, usize, usize) {
        (self.lnum, self.bol, self.offset)
    }
}

impl FilePos for FilePosLarge {
    #[inline]
    fn offset(&self) -> usize {
        self.offset
    }

    #[inline]
    fn line_column_beg(&self) -> (usize, usize, usize) {
        (self.lnum, self.offset - self.bol, self.bol)
    }
}

impl Ord for FilePosLarge {
    fn cmp(&self, other: &FilePosLarge) -> std::cmp::Ordering {
        self.offset().cmp(&other.offset())
    }
}

impl PartialOrd for FilePosLarge {
    fn partial_cmp(&self, other: &FilePosLarge) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

impl From<FilePosSmall> for FilePosLarge {
    fn from(pos: FilePosSmall) -> Self {
        let (lnum, bol, offset) = pos.line_beg_offset();
        Self::from_lnum_bol_offset(lnum, bol, offset)
    }
}

impl ToOcamlRep for FilePosLarge {
    fn to_ocamlrep<'a, A: ocamlrep::Allocator>(&'a self, alloc: &'a A) -> ocamlrep::Value<'a> {
        let mut block = alloc.block_with_size(3);
        alloc.set_field(&mut block, 0, alloc.add(&self.lnum));
        alloc.set_field(&mut block, 1, alloc.add(&self.bol));
        alloc.set_field(&mut block, 2, alloc.add_copy(self.offset as isize));
        block.build()
    }
}

impl FromOcamlRep for FilePosLarge {
    fn from_ocamlrep(value: ocamlrep::Value<'_>) -> Result<Self, ocamlrep::FromError> {
        let block = ocamlrep::from::expect_tuple(value, 3)?;
        let lnum = ocamlrep::from::field(block, 0)?;
        let bol = ocamlrep::from::field(block, 1)?;
        let offset: isize = ocamlrep::from::field(block, 2)?;
        Ok(Self {
            lnum,
            bol,
            offset: offset as usize,
        })
    }
}

impl<'a> FromOcamlRepIn<'a> for FilePosLarge {
    fn from_ocamlrep_in(
        value: ocamlrep::Value<'_>,
        _alloc: &'a ocamlrep::Bump,
    ) -> Result<Self, ocamlrep::FromError> {
        Self::from_ocamlrep(value)
    }
}

File: hphp/hack/src/utils/file_pos_large.ml
(*
 * Copyright (c) 2015, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the "hack" directory of this source tree.
 *
 *)

open Hh_prelude

(* See documentation in mli file *)
type t = {
  pos_lnum: int;
  pos_bol: int;
  pos_offset: int;
}
[@@deriving eq, hash, ord]

let pp fmt pos =
  Format.pp_print_int fmt pos.pos_lnum;
  Format.pp_print_string fmt ":";
  Format.pp_print_int fmt (pos.pos_offset - pos.pos_bol + 1)

let show pos = Format.asprintf "%a" pp pos

let compare : t -> t -> int = compare

let dummy = { pos_lnum = 0; pos_bol = 0; pos_offset = -1 }

let is_dummy = equal dummy

let beg_of_file = { pos_lnum = 1; pos_bol = 0; pos_offset = 0 }

(* constructors *)

let of_line_column_offset ~line ~column ~offset =
  { pos_lnum = line; pos_bol = offset - column; pos_offset = offset }

let of_lexing_pos lp =
  {
    pos_lnum = lp.Lexing.pos_lnum;
    pos_bol = lp.Lexing.pos_bol;
    pos_offset = lp.Lexing.pos_cnum;
  }

let of_lnum_bol_offset ~pos_lnum ~pos_bol ~pos_offset =
  { pos_lnum; pos_bol; pos_offset }

(* accessors *)

let offset t = t.pos_offset

let line t = t.pos_lnum

let column t = t.pos_offset - t.pos_bol

let beg_of_line t = t.pos_bol

let set_column c p =
  { pos_lnum = p.pos_lnum; pos_bol = p.pos_bol; pos_offset = p.pos_bol + c }

let line_beg t = (t.pos_lnum, t.pos_bol)

let line_column t = (t.pos_lnum, t.pos_offset - t.pos_bol)

let line_column_beg t = (t.pos_lnum, t.pos_offset - t.pos_bol, t.pos_bol)

let line_column_offset t = (t.pos_lnum, t.pos_offset - t.pos_bol, t.pos_offset)

let line_beg_offset t = (t.pos_lnum, t.pos_bol, t.pos_offset)

let to_lexing_pos pos_fname t =
  {
    Lexing.pos_fname;
    Lexing.pos_lnum = t.pos_lnum;
    Lexing.pos_bol = t.pos_bol;
    Lexing.pos_cnum = t.pos_offset;
  }
